services/typeddata/sign.go

```diff
@@ -0,0 +1,51 @@
+package typeddata
+
+import (
+	"crypto/ecdsa"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+const (
+	chainIDKey = "chainId"
+)
+
+var (
+	// x19 to avoid collision with rlp encode. x01 version byte defined in EIP-191
+	messagePadding = []byte{0x19, 0x01}
+)
+
+func hashToSign(typed TypedData) (rst common.Hash, err error) {
+	domainSeparator, err := encodeData(eip712Domain, typed.Domain, typed.Types)
+	if err != nil {
+		return rst, err
+	}
+	primary, err := encodeData(typed.PrimaryType, typed.Message, typed.Types)
+	if err != nil {
+		return rst, err
+	}
+	return crypto.Keccak256Hash(messagePadding, domainSeparator[:], primary[:]), nil
+}
+
+// Sign TypedData with a given private key. Verify that chainId in the typed data matches currently selected chain.
+func Sign(typed TypedData, prv *ecdsa.PrivateKey, chain *big.Int) ([]byte, error) {
+	if _, exist := typed.Domain[chainIDKey]; !exist {
```

Comment:
I would suggest to move this block into a separate method of `TypedData` called `ValidateChainID` or similar.

So it could be just:

```go
    err := typed.validateChainID(chain)
    if err != nil {
        return nil, fmt.Errorf("chain id: %v", err)
    }
```

Reply:
it makes TypedData depend on hardcoded chainIDkey. not sure that it is better. but cleaner